src/GraphQL/Federation/Extensions/FederationGraphTypeExtensions.cs

```diff
@@ -0,0 +1,110 @@
+            .Resolve(FederationQuerySchemaNodeVisitor.ResolveEntities);
+    }
+
+    /// <summary>
+    /// Adds "@key" directive.
+    /// </summary>
+    public static void Key(this IObjectGraphType graphType, string[] fields, bool resolvable = true) =>
+        graphType.Key(string.Join(" ", fields.Select(x => x.ToCamelCase())), resolvable);
+
+    /// <summary>
+    /// Adds "@key" directive.
+    /// </summary>
+    public static void Key(this IObjectGraphType graphType, string fields, bool resolvable = true)
+    {
+        var astMetadata = graphType.BuildAstMetadata();
+        var directive = new GraphQLDirective { Name = new(KEY_DIRECTIVE) };
+        directive.AddFieldsArgument(fields.ToCamelCase());
+        directive.AddResolvableArgument(resolvable);
+        astMetadata!.Directives!.Items.Add(directive);
+    }
+
+    /// <summary>
+    /// Adds "@shareable" directive.
+    /// </summary>
+    public static void Shareable(this IGraphType graphType)
+    {
+        if (graphType.IsInputType())
+            throw new ArgumentOutOfRangeException(nameof(graphType), graphType, "Input types are not supported.");
+        var astMetadata = graphType.BuildAstMetadata();
+        var directive = new GraphQLDirective { Name = new(SHAREABLE_DIRECTIVE) };
+        astMetadata!.Directives!.Items.Add(directive);
+    }
+
+    /// <summary>
+    /// Adds "@inaccessible" directive.
+    /// </summary>
+    public static void Inaccessible(this IGraphType graphType)
+    {
+        if (graphType.IsInputType())
+            throw new ArgumentOutOfRangeException(nameof(graphType), graphType, "Input types are not supported.");
+        var astMetadata = graphType.BuildAstMetadata();
+        var directive = new GraphQLDirective { Name = new(INACCESSIBLE_DIRECTIVE) };
+        astMetadata!.Directives!.Items.Add(directive);
+    }
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, TSourceType> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new FederationResolver<TSourceType>(resolver);
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, Task<TSourceType>> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new AsyncFederationResolver<TSourceType>(resolver);
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, IDataLoaderResult<TSourceType>> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new DataLoaderFederationResolver<TSourceType>(resolver);
```

Comment:
See here, where we include extension methods for up to 3 levels of data loaders, for both synchronous and asynchronous resolvers:
- https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL.DataLoader/Extensions/DataLoaderExtensions.cs#L173

Reply:
It's been too long but as I remember I had one class initially and then switched to this becausew it was making something else easier. I'll try again.

Reply:
I simplified it to some extend. I got rid of multiple resolver classes and kept only one `FuncFederationResolver<T>`. Made both `IFederationResolver` and `FuncFederationResolver<T>` and added extensions to pass any `IFederationResolver`. What I wasn't able to do is making it `ValueTask<object>` instead of just `object` (see above).